[MRG] make transformer_from_metric more robust

wdevazelhes opened the pull request:
Fixes #175 

This PR fixes the problem we had with `transformer_from_metric`, by checking if the matrix is diagonal depending on the relative gap between the maximum absolute value of the outer diagonal coeffs, and the minimum absolute value of the diagonal coefficients, rather than what was done before.

It is also more robust for detecting whether to use Cholesky or the eigendecomposition: instead of computing the determinant to check if the matrix is not definite, it tries to do Cholesky and if an error is returned it does the eigendecomposition. Indeed, maybe the determinant can be not null if there is a very small eigenvalue (`v`=1e-5) (that should be considered as null), but a lot of big eigenvalues (see one example in the test)

It also raises an error message when the input is not symmetric and a warning when switching from the Cholesky decomposition to the eigendecompostion if the Cholesky decomposition is not doable.


TODO:
- [x] to make a good test, check the value of `v` for which cholesky will fail, and from it build the example where the determinant is not "np.close" to zero

wdevazelhes commented:
Now that I think about it, maybe the warning is not a good thing to throw ? Maybe it would be better a verbose flag to decide whether to print the warning (which would not be anymore a warning but a regular print) ? What do you think ?

wdevazelhes commented:
I agree, I just pushed a more recent version that takes into account these remarks

bellet commented on metric_learn/_util.py:
```diff
@@ -324,7 +326,32 @@ def check_collapsed_pairs(pairs):
+
+  tol : float, optional
+    Negative eigenvalues above - tol are considered zero. If
+    tol is None, and w are `metric`'s eigenvalues, and eps is the
```
"and w are `metric`'s eigenvalues" is not needed

wdevazelhes replied:
That's right, done

bellet commented on metric_learn/_util.py:
```diff
@@ -324,7 +326,32 @@ def check_collapsed_pairs(pairs):
+  """
+  if tol is None:
+    tol = w.max() * len(w) * np.finfo(w.dtype).eps
+  if any(w[w < 0] < - tol):
```
`w<0` not needed

bellet replied:
although it actually prevents that anything is done in case the provided `tol` is negative (which is not supposed to be the case according to the docstring but is not checked)

wdevazelhes replied:
That's right, I'll let it be w < - tol and add a quick test to check that tol is positive

bellet commented on metric_learn/_util.py:
```diff
@@ -337,18 +364,34 @@ def transformer_from_metric(metric):
+
+  tol : positive float, optional
+    Eigenvalues of `metric` between 0 and - tol are considered zero. If tol is
+    None, and w are `metric`'s eigenvalues, and eps is the epsilon value for
```
same as above

wdevazelhes replied:
How to explain how the tolerance is set here then without talking about the eigenvalues ? Though maybe an explanation talking about the highest eigenvalue could be better ? Like:
```
    Eigenvalues of `metric` between 0 and - tol are considered zero. If tol is
    None, and w_max is `metric`'s largest eigenvalue, and eps is the epsilon
    value for datatype of w, then tol is set to w_max * metric.shape[0] * eps.
```

bellet replied:
nevermind, indeed here we need to mention eigenvalues
the suggested reformulation looks good

bellet commented on test/test_transformer_metric_conversion.py:
```diff
@@ -76,6 +81,102 @@ def test_mlkr(self):
+    P = ortho_group.rvs(7, random_state=rng)
+    M = P.dot(D).dot(P.T)
+    with pytest.raises(ValueError) as raised_error:
+      transformer_from_metric(M)
```
here you could also easily test that the error is raised in the diagonal case, for instance when calling `transformer_from_metric(D)`

wdevazelhes replied:
I agree, will do

wdevazelhes commented:
Thanks for the review @bellet, I addressed all the comments, there's just ![this one](https://github.com/metric-learn/metric-learn/pull/191#discussion_r275857008) that needs to be resolved

bellet commented:
LGTM!
